File: src/reconfigurable_client.cpp
#include "uwds/reconfigurable_client.h"

using namespace std;
using namespace std_msgs;
using namespace uwds_msgs;

namespace uwds
{
  void ReconfigurableClient::onInit()
  {
    UwdsClientNodelet::onInit();
    pnh_->param<bool>("use_single_input", use_single_input_, false);
    string default_inputs;
    vector<string> input_worlds;
    pnh_->param<string>("default_inputs", default_inputs, "");
    boost::split(input_worlds, default_inputs, boost::is_any_of(" "));
    reconfigure(input_worlds);
    reconfigure_service_server_ = nh_->advertiseService(ctx_->name()+"/reconfigure_inputs", &ReconfigurableClient::reconfigureInputs, this);
    list_inputs_service_server_ = nh_->advertiseService(ctx_->name()+"/list_inputs", &ReconfigurableClient::listInputs, this);
  }

  void ReconfigurableClient::reconfigure(vector<string> inputs)
  {
    if (inputs.size() > 1 and use_single_input_)
    {
      throw std::runtime_error("Multiple inputs provided while 'use_single_input' activated.");
    }
    ctx_->worlds().close();
    onReconfigure(inputs);
    input_worlds_.clear();

    for(const string input : inputs)
    {
       ctx_->worlds()[input].connect(bind(&ReconfigurableClient::onChanges, this, _1, _2, _3));
       Invalidations invalidations;
       auto& scene = ctx_->worlds()[input].scene();
       auto& timeline = ctx_->worlds()[input].timeline();
       auto& meshes = ctx_->worlds()[input].meshes();
       for (const auto& node : scene.nodes())
          if (node->name !="root")
            invalidations.node_ids_updated.push_back(node->id);
       for (const auto& situation : timeline.situations())
          invalidations.situation_ids_updated.push_back(situation->id);
       for (const auto& mesh : meshes.meshes())
          invalidations.mesh_ids_updated.push_back(mesh->id);
       Header header;
       header.frame_id = global_frame_id_;
       header.stamp = ros::Time::now();
       onChanges(input, header, invalidations);
    }
    input_worlds_ = inputs;
  }

  bool ReconfigurableClient::reconfigureInputs(ReconfigureInputs::Request& req,
                                               ReconfigureInputs::Response& res)
  {
    if(verbose_)NODELET_INFO("[%s::reconfigure] Service '~reconfigure_inputs' requested", ctx_->name().c_str());
    try
    {
      reconfigure(req.inputs);
      res.success = true;
    } catch(const std::exception& e) {
      res.success = false;
      res.error = e.what();
    }
    return true;
  }

  bool ReconfigurableClient::listInputs(List::Request& req,
                                        List::Response& res)
  {
    res.list = inputsWorlds();
    res.success = true;
    return true;
  }
}

File: src/pyuwds/reconfigurable_client.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import rospy
from std_msgs.msg import Header
from uwds_msgs.msg import Invalidations
from uwds_msgs.srv import ReconfigureInputs, List
from pyuwds.uwds_client import UwdsClient


class ReconfigurableClient(UwdsClient):
    """
    The Underworlds client

    @type self.node_name: string
    @param self.node_name: The client name
    """
    def __init__(self, client_name, client_type):
        """
        The Underworlds client

        @type self.node_name: string
        @param self.node_name: The client name
        """
        super(ReconfigurableClient, self).__init__(client_name, client_type)
        input_worlds = rospy.get_param("~default_inputs", "")
        self.__use_single_input = rospy.get_param("~use_single_input", False)
        self.input_worlds = input_worlds.split(" ")
        self.reconfigure(input_worlds.split(" "))
        self.__reconfigure_service_server = rospy.Service(client_name+"/reconfigure_inputs", ReconfigureInputs, self.reconfigureInputs)
        self.__list_inputs_service_server = rospy.Service(client_name+"/list_inputs", List, self.listInputs)
        rospy.loginfo("[{}::init] Underworlds client ready !".format(self.ctx.name()))

    def reconfigure(self, inputs):
        if len(inputs) > 1 and self.__use_single_input:
            raise RuntimeError("Multiple inputs provided while 'use_single_input' activated.")
        self.ctx.worlds().close()
        self.onReconfigure(inputs)
        for input in inputs:
            self.ctx.worlds()[input].connect(self.onChanges)
            invalidations = Invalidations()
            scene = self.ctx.worlds()[input].scene()
            timeline = self.ctx.worlds()[input].timeline()
            meshes = self.ctx.worlds()[input].meshes()
            for node in scene.nodes():
                invalidations.node_ids_updated.append(node.id)
            for situation in timeline.situations():
                invalidations.situation_ids_updated.append(situation.id)
            for mesh in meshes:
                invalidations.mesh_ids_updated.append(mesh.id)
            header = Header()
            header.stamp = rospy.Time.now()
            self.onChanges(input, header, invalidations)
        self.inputs_worlds = inputs

    def reconfigureInputs(self, req):
        try:
            self.reconfigure(req.inputs)
            return True, ""
        except Exception as e:
            return False, str(e)

    def listInputs(self, req):
        try:
            return self.input_worlds, True, ""
        except Exception as e:
            return [], False, str(e)

    def onChanges(self, world_name, header, invalidations):
        raise NotImplementedError

    def onReconfigure(self, input_worlds):
        raise NotImplementedError
